packages/eslint-plugin/tests/rules/strict-type-predicates.test.ts

```diff
@@ -0,0 +1,626 @@
+}
+
+// typeof object
+typeof get<{}>() === "object";
+
+// === null / undefined
+// get<number | null>() === null;
+// get<number | undefined>() === undefined;
+// get<string | null | undefined>() == null;
+// get<string | null | undefined>() != undefined;
+
+// negation
+get<number | null>() !== null;
+get<number | undefined>() !== undefined;
+get<unknown>() !== null;
+get<unknown>() !== undefined;
+
+// type parameters
+{
+  function f<T>(t: T) {
+    typeof t === "boolean";
+  }
+
+  // TODO: Would be nice to catch this.
+  function g<T extends string>(t: T) {
+    typeof t === "boolean";
+  }
+
+  function f<T>(t: T) {
+    typeof t === "boolean";
+  }
+}
+
+// Detects bad typeof
+{
+  typeof get<string | number>() === \`string\`;
+  let a: string, b: string;
+  typeof a === typeof b;
+  typeof a === b;
+  a === typeof b;
+}
+
+// unknown
+typeof get<unknown>() === "undefined";
+typeof get<unknown>() === "boolean";
+typeof get<unknown>() === "number";
+typeof get<unknown>() === "string";
+typeof get<unknown>() === "symbol";
+typeof get<unknown>() === "function";
+typeof get<unknown>() === "object";
+"string" === typeof get<unknown>();
+undefined === get<unknown>();
+
+// other
+{
+  const body: unknown = 'test';
+  if (typeof body === 'object')
+    console.log('a');
+
+  let test: unknown = undefined;
+  if (test !== undefined)
+    console.log('b');
+}
+    `,
+  ],
+
+  invalid: [
+    // typeof undefined
+    {
+      code: `
+declare function get<T>(): T;
+typeof get<boolean>() === "undefined";`,
+      errors: [
+        {
+          messageId: 'expressionAlwaysFalse',
+          line: 3,
+          column: 1,
+        },
+      ],
+    },
+    {
+      code: `
+declare function get<T>(): T;
+typeof get<void>() === "undefined";`,
+      errors: [
+        {
+          messageId: 'expressionAlwaysTrue',
+          line: 3,
+          column: 1,
+        },
+      ],
+    },
+    {
+      code: `
+declare function get<T>(): T;
+// 'undefined' is not assignable to '{}'
+typeof get<{}>() === "undefined";`,
+      errors: [
+        {
+          messageId: 'expressionAlwaysFalse',
+          line: 4,
+          column: 1,
+        },
+      ],
+    },
+
+    // typeof boolean
+    {
+      code: `
+declare function get<T>(): T;
+typeof get<boolean>() === "boolean";`,
+      errors: [
+        {
+          messageId: 'expressionAlwaysTrue',
+          line: 3,
+          column: 1,
+        },
+      ],
+    },
+    {
+      code: `
+declare function get<T>(): T;
+typeof get<Boolean>() === "boolean";`,
+      errors: [
+        {
+          messageId: 'expressionAlwaysFalse',
+          line: 3,
+          column: 1,
+        },
+      ],
+    },
+
+    // typeof number
+    {
+      code: `
+declare function get<T>(): T;
+enum E {}
+typeof get<E>() === "number";`,
+      errors: [
+        {
+          messageId: 'expressionAlwaysTrue',
+          line: 4,
+          column: 1,
+        },
+      ],
+    },
```

Comment:
for completeness, would be good to get a string enum test `enum E { a = 'aaa' }`.